### ISSUE TITLE
Upgrader creates a branch with explanatory commits

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-upgrader#92]
## Problem

As [discussed elsewhere](https://github.com/silverstripe/silverstripe-upgrader/issues/72#issuecomment-392728783), the upgrader command can be a bit of a black box. It's important that developers know what upgrader has done, and why. It's also important that this information is put into a place a little more persistent than a CLI output - you'll likely want to refer to it for the duration of an upgrade project.

Sometimes, it's just to know what's happened. Sometimes it is to put more human attention into specific pieces. For example, the rewriting of string-references to classnames is prone to false-positives (e.g. a field label that is now "SilverStripe\Email\Email").

## Recommended solution

Let's assume that most people are managing their code in a git codebase. Given that, in my view, the best way of presenting this would be a git branch with a series of commits for the various transformations that the upgrader has performed.

As well as a brief explanation of what it's done, the commit message could also contain a deep-link to the docs that explain what has happened.

In practise, the developer can then add to that branch with their own manually created patches. After they're done, they can squash the branch into master, or retain the full commit history, at their preference.

## Implementation details

 * This behaviour could be triggered with a CLI option such as `--create-branch ss4-upgrade` on the all-in-one mode, where ss4-upgrade is the name of the branch to be created.

 * On a per-step basis, we might have a `--create-commit "(commit message)"` option, which is mainly designed to be populated by the all-in-one command.

 * With this in place, it would be really helpful to separate out the string rewriting part of the class renamer from the regular classname & yaml rewriting. Maybe this could be more generally described as "safe rewrites" and "risky rewrites". This will let me put more review attention into the risky rewrites.

**Comment by chillu** [comment]
This sounds like a fairly time intensive way to increase comms. Wouldn't it be nearly as valueable to just explain this in each step of the upgrader CLI output, and suggest that it's a good idea to commit the changes at this point before continuing?

**Comment by sminnee** [comment]
Yeah I guess you could ask a human to do this automatable process themselves. But that seems counter to the whole purpose of the upgrader tool. ;)

I raised this mainly because on the upgrades that I've tried myself, I've typically done this manually (e.g. starting with a commit that captures a phpcfg execution)

Are you saying "sounds like a nice idea but not on the critical path" or "sounds like a bad idea"?

**Comment by chillu** [comment]
I'm just arguing that it's not on the critical path, we need to wrap up the upgrader work and move on to other things, getting into diminishing returns compared to other product areas.

**Comment by sminnee** [comment]
OK I might save this for a hackday then. ;-)

**Comment by maxime-rainville** [comment]
In case we pick this up later on ... I've coded the all-in-one command with the expectation that it could be run over and over again (e.g.: Debugging one error at a time.) I'm not sure what the best way to handle commits and branching in this scenario would be.

**Comment by sminnee** [comment]
Good to know. Perhaps something like only triggering a commit after a step completes successfully?

**Comment by sunnysideup** [comment]
You can look at the upgrader I am working on (which basically wraps around silverstripe/upgrader and creates a branch, runs the commands, and then commits it in branch: https://github.com/sunnysideup/silverstripe-upgrade_to_silverstripe_4
